src/AppContainer.dev.js

```diff
@@ -0,0 +1,109 @@
+'use strict';
+
+const React = require('react');
+const deepForceUpdate = require('react-deep-force-update');
+const Redbox = require('redbox-react');
+const { Component } = React;
+
+// Feature check for the createElement() patch.
+// If createElement() was patched, types with
+// the same __source will resolve to the same type.
+let wasCreateElementPatched = false;
+const A = () => {};
+A.__source = { fileName: 'fake', localName: 'fake' }
+const B = () => {};
+B.__source = { fileName: 'fake', localName: 'fake' }
+if (<A />.type === <B />.type) {
+  wasCreateElementPatched = true;
+}
+
+class AppContainer extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { error: null };
+  }
+
+  componentDidMount() {
+    if (!wasCreateElementPatched) {
+      console.error(
+        'React Hot Loader: It appears that "react-hot-loader/patch" ' +
+        'did not run immediately before the app started. Make sure that it ' +
+        'runs before any other code. For example, if you use Webpack, ' +
+        'you can add "react-hot-loader/patch" as the very first item to the ' +
+        '"entry" array in its config. Alternatively, you can add ' +
+        'require("react-hot-loader/patch") as the very first line ' +
+        'in the application code, before any other imports.'
+      );
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // Hot reload is happening.
+    // Retry rendering!
+    this.setState({
+      error: null
+    });
+  }
+
+  componentDidUpdate(prevProps) {
+    // Hot reload has finished.
+    // Force-update the whole tree, including
+    // components that refuse to update.
+    if (prevProps !== this.props) {
+      deepForceUpdate(this);
```

Comment:
@gaearon calling `deepForceUpdate` in `componentWillUpdate` makes the children component re-render twice every time it changes, one due to the regular react lifecycle, another due to the `deepForceUpdate`. If we moved `deepForceUpdate` to `componentWillReceiveProps` it would re-render only once. Do you see any downside?